cdisutils/storage3.py

```diff
@@ -0,0 +1,668 @@
+                                key_data.append(line_data)
+                    else:
+                        # ok, let's see if we can be smart here
+                        #if not header:
+                        #    remaining_chars = set([c for c in line if not c.isalnum()])
+                        skipped_lines += 1
+
+        self.log.info('%d lines in file, %d processed',
+                      len(file_data.split('\n')), len(key_data))
+        return key_data
+
+    def checksum_s3_key(self,
+                        url=None):
+        ''' Get the checksum of an s3 object '''
+        result = {
+            'transfer_time': 0,
+            'bytes_transferred': 0
+        }
+        md5sum = hashlib.md5()
+        sha = hashlib.sha256()
+        retries = 0
+        result['start_time'] = time.time()
+        running = False
+        total_transfer = 0
+        file_key_info = self.get_url(url=url)
+        if file_key_info:
+            file_key = file_key_info.get('Body', None)
+            file_key_size = file_key_info.get('ContentLength', None)
+            running = True
+            #file_key.BufferSize = self.chunk_size
+        else:
+            self.log.warning("Unable to get %s ", url)
+
+        while running:
+            try:
+                chunk = self.download_object_part(key=file_key)
+            except ClientError as exception:
+                if chunk:
+                    if retries > 10:
+                        self.log.error("Error reading: %s", exception)
+                        break
+                    else:
+                        retries += 1
+                        self.log.error("Error reading: %s retry %d", exception, retries)
+                        time.sleep(2)
+                else:
+                    self.log.error("Error reading %s, got %d bytes",
+                                   exception, len(chunk))
+                    total_transfer += len(chunk)
+                    md5sum.update(chunk)
+                    sha.update(chunk)
+                    retries = 0
+            else:
+                result['bytes_transferred'] += len(chunk)
+                if (len(chunk) < self.chunk_size) and\
+                    (result['bytes_transferred'] >= file_key_size):
+
+                    running = False
+
+                if file_key_size > 0:
+                    sys.stdout.write("{:6.02f}%\r".format(
+                        (float(result['bytes_transferred']) /
+                        float(file_key_size) * 100.0)))
+                else:
+                    sys.stdout.write("0.00%%\r")
+                sys.stdout.flush()
+                md5sum.update(chunk)
+                sha.update(chunk)
+                retries = 0
+
+        result['transfer_time'] = time.time() - result['start_time']
+        result['md5_sum'] = md5sum.hexdigest()
+        result['sha256_sum'] = sha.hexdigest()
+        return result
```

Comment:
668 lines of untested code. :(

Reply:
I will file a ticket to add tests for this. It's getting to be a little outside the scope of my current task.